move 'floating' vCPU code to the `vcpu` struct

acatangiu opened the pull request:
Move the main loop being run by vCPU threads from the generic `vmm/src/lib.rs` inside the implementation of the existing `vcpu` struct defined in `vmm/src/vstate.rs`.

Fixes https://github.com/firecracker-microvm/firecracker/issues/1002.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

alexandruag commented on vmm/src/vstate.rs:
```diff
@@ -275,13 +290,139 @@ impl Vcpu {
+        }
+    }
+
+    fn run_guest(&self) -> Result<VcpuExit> {
```
Is this really needed as an individual method? It's only used in two places, and it looks likes we could inline it by hand in both of them.

acatangiu replied:
Sounds good.

acatangiu replied:
done

alexandruag commented on vmm/src/lib.rs:
```diff
@@ -1221,18 +1111,15 @@ impl Vmm {
     }
 
     fn start_microvm(&mut self) -> std::result::Result<VmmData, VmmActionError> {
-        START_INSTANCE_REQUEST_CPU_TS.store(now_cputime_us() as usize, Ordering::Release);
```
Why can't we keep `START_INSTANCE_REQUEST_CPU_TS` and `START_INSTANCE_REQUEST_TS` as they were before? They're still accessible from `vstate`, we wouldn't have to pass them around, and they can also be used from other places should that ever matter.

acatangiu replied:
I was aiming for getting rid of global state since it's not in any way necessary for these to be global.

There should never be *other places* that they'd be needed in.

I think it's more in lines with the general model of Firecracker's codebase to *move* them to where they're needed rather than keeping them as global state. Logical flow is that the `vmm` thread initializes the TS and then it's *moved/consumed* by the `vcpu`s who are then the sole users of the TS local copies.

acatangiu replied:
Insisting on not using global vars especially when it can be so easily avoided.

alexandruag commented on vmm/src/vstate.rs:
```diff
@@ -192,6 +197,8 @@ pub struct Vcpu {
     cpuid: CpuId,
     fd: VcpuFd,
     id: u8,
+    start_instance_request_cpu_ts: u64,
```
I think these two look really out of place as members of `Vcpu`. If there's some compelling reason against leaving them as globals in `lib.rs` as before, we should pass them as parameters to `run()` or something like that.

acatangiu replied:
I agree that they look kinda out of place inside the `Vcpu`, but passing them as parameters involves loading them as parameters for quite a number of functions going down the stack to where they're actually used.

I'm working on some other code changes in this area and the only *nice* ways are to have them as:
- local object properties, or
- process-level global state.

I can argue that even though they *look out of place*, they are still properties of the `Vcpu` as they hold the TS for when *that `Vcpu` object instance* was created and are used to provide functionality for *that particular `Vcpu` instance*.

acatangiu replied:
Changed to creation timestamp as member of vcpu

alexandruag commented on vmm/src/vstate.rs:
```diff
@@ -275,13 +290,139 @@ impl Vcpu {
+        mmio_bus: devices::Bus,
+        vcpu_exit_evt: EventFd,
+    ) {
+        unsafe {
```
Like you also rightfully mentioned, this part of the code is written as if we're setting the signal handler per thread, when it's actually set for the entire process. It looks like we can install the handler in the Vmm before running the vcpus.

acatangiu replied:
Yes, I have that change in a following PR, but I agree it might be better suited in this one.

acatangiu replied:
done

alexandruag commented on vmm/src/vstate.rs:
```diff
@@ -275,13 +290,139 @@ impl Vcpu {
         self.fd.run().map_err(Error::VcpuRun)
     }
+
+    fn log_boot_time(&self) {
```
This looks more at home as a global function in the parent module (which can still be invoked in this one), especially if keep the original form of `START_INSTANCE_REQUEST_CPU_TS` and `START_INSTANCE_REQUEST_TS`.

acatangiu replied:
Makes sense. I'm still keen on not using global state if not necessary, but either way this would be more at home as a global function in the parent module. Will change.

acatangiu replied:
done

acatangiu commented:
@serban300 @alexandruag I've addressed your comments.

I've also added tests.
I would've liked to be able to test all exit reasons in `run_emulation()` but because `Result<VcpuExit>` keeps a reference of `self`, I wasn't able to refactor in such a way to make that possible.

I have some further pending work in this area, and I'm hoping with those changes we'll be able to decouple the emulation match and the actual KVM_RUN ioctl so that we can unit-test all the supported `VcpuExit` reasons.

Please review as this is blocking further work for me.

andreeaflorescu commented:
I want to propose to squash the two commits and as a general rule to update the coverage in the commit that adds unit tests. My concern is that we will end up with a lot of commits that only update the coverage and thus making the git history a bit strange.

acatangiu commented:
@andreeaflorescu done. But now I lost @serban300's approve :disappointed:

andreeaflorescu commented:
He's a nice guy. Bet he'll re-approve it!

acatangiu commented:
@andreeaflorescu He's on vacation. Did you review the whole thing or you just had that drive-by comment?

serban300 commented:
> @andreeaflorescu He's on vacation. Did you review the whole thing or you just had that drive-by comment?

I'm not on vacation. I will take another look on this

acatangiu commented:
@firecracker-microvm/compute-capsule I've rebased on top of the merged seccomp PR and resolved conflicts. Please review.

aghecenco commented on vmm/src/lib.rs:
```diff
@@ -988,14 +989,26 @@ impl Vmm {
             .ok_or(StartMicrovmError::VcpusNotConfigured)?;
         let mut vcpus = Vec::with_capacity(vcpu_count as usize);
+
+        // mmio_device_manager is instantiated in init_devices, which is called before
```
nit: enclose mmio_device_manager, init_devices and start_vcpus in backtics

acatangiu replied:
done

aghecenco commented on vmm/src/lib.rs:
```diff
@@ -1248,18 +1134,17 @@ impl Vmm {
             ));
         }
+        let request_ts = TimestampUs {
+            time_us: (chrono::Utc::now().timestamp_nanos() / 1000) as u64,
```
nit: maybe add an inline shorthand for `(chrono::Utc::now().timestamp_nanos() / 1000) as u64`, like `now_cputime_us`.

acatangiu replied:
This is real time not cpu time, but I added an inline function for it.

aghecenco commented on vmm/src/lib.rs:
```diff
@@ -1954,6 +1839,21 @@ impl Vmm {
         Ok(())
     }
+
+    fn log_boot_time(t0_ts: TimestampUs) {
```
Why not borrow `TimestampUs`?

acatangiu replied:
done

aghecenco commented on vmm/src/lib.rs:
```diff
@@ -1954,6 +1839,21 @@ impl Vmm {
+
+        let boot_time_us = now_us - t0_ts.time_us;
+        let boot_time_cpu_us = now_cpu_us - t0_ts.cputime_us;
+        warn!(
```
Why `warn!`?

acatangiu replied:
That was the default in integ tests :stuck_out_tongue:, I've changed both to `info`.

aghecenco commented on vmm/src/vstate.rs:
```diff
@@ -261,19 +278,124 @@ impl Vcpu {
+                    libc::EAGAIN | libc::EINTR => Ok(()),
+                    _ => {
+                        METRICS.vcpu.failures.inc();
+                        error!("Failure during vcpu run: {:?}", e);
```
`e` is an `std::io::Error`, please use the `{}` formatter instead.

acatangiu replied:
done

acatangiu commented on vmm/src/lib.rs:
```diff
@@ -3031,6 +2932,39 @@ mod tests {
+        }
+
+        // Validate logging the boot time works.
+        Vmm::log_boot_time(TimestampUs {
```
done
